rpcs3/Emu/Cell/Modules/sys_net.cpp

```diff
@@ -150,27 +145,72 @@ namespace sys_net
 		return g_tls_net_data.ref(&_tls_data_t::_h_errno);
 	}
 
+	// Error helper functions
+	s32 get_last_error()
+	{
+		// Convert the error code for socket functions to a one for sys_net
+#ifdef _WIN32
+		switch (WSAGetLastError())
+		{
+		case WSAEWOULDBLOCK:
+			return SYS_NET_EWOULDBLOCK;
+
+		default:
+			throw EXCEPTION("Unknown Win32 socket error: %d", WSAGetLastError());
+		}
+#else
+		switch (errno)
+		{
+		case EWOULDBLOCK:
+			return SYS_NET_EWOULDBLOCK;
+
+		default:
+			throw EXCEPTION("Unknown Unix socket error: %d", errno);
+		}
+
+		return errno;
+#endif
+	}
+
 	// Functions
 	s32 accept(s32 s, vm::ptr<sockaddr> addr, vm::ptr<u32> paddrlen)
 	{
 		libnet.warning("accept(s=%d, family=*0x%x, paddrlen=*0x%x)", s, addr, paddrlen);
 		s = g_socketMap[s];
 
-		if (!addr) {
-			int ret = ::accept(s, nullptr, nullptr);
-			get_errno() = getLastError();
-			return ret;
+		s32 ret;
+
+		if (!addr)
+		{
+			ret = ::accept(s, nullptr, nullptr);
+
+			if (ret < 0)
+			{
+				libnet.error("accept(): error %d", get_errno() = get_last_error());
+				return -1;
+			}
 		}
-		else {
+		else
+		{
 			::sockaddr _addr;
-			memcpy(&_addr, addr.get_ptr(), sizeof(::sockaddr));
-			_addr.sa_family = addr->sa_family;
-			::socklen_t _paddrlen;
-			s32 ret = ::accept(s, &_addr, &_paddrlen);
+			::socklen_t _paddrlen = 16;
+
+			ret = ::accept(s, &_addr, &_paddrlen);
+
+			if (ret < 0)
+			{
+				libnet.error("accept(): error %d", get_errno() = get_last_error());
+				return -1;
+			}
+
 			*paddrlen = _paddrlen;
-			get_errno() = getLastError();
-			return ret;
+			addr->sa_len = _paddrlen;
+			addr->sa_family = _addr.sa_family;
+			memcpy(addr->sa_data, _addr.sa_data, addr->sa_len - 2);
 		}
+
+		g_socketMap.push_back(ret);
```

Comment:
Yes and I was about to suggest a solution from my git stash. Actually I improved IdManager recently a bit, so it can handle these sockets easily.